src/aggregators/ssajump.jl

```diff
@@ -119,16 +115,68 @@ end
 
 # Update state based on the p.next_jump
 @inline function update_state!(p::AbstractSSAJumpAggregator, integrator, u)
-    num_ma_rates = get_num_majumps(p.ma_jumps)
-    if p.next_jump <= num_ma_rates # is next jump a mass action jump
+    @unpack ma_jumps, next_jump = p
+    num_ma_rates = get_num_majumps(ma_jumps)
+    if next_jump <= num_ma_rates # is next jump a mass action jump
         if u isa SVector
-            integrator.u = executerx(u, p.next_jump, p.ma_jumps)
+            integrator.u = executerx(u, next_jump, ma_jumps)
         else
-            @inbounds executerx!(u, p.next_jump, p.ma_jumps)
+            @inbounds executerx!(u, next_jump, ma_jumps)
         end
     else
-        idx = p.next_jump - num_ma_rates
-        @inbounds p.affects![idx](integrator)        
+        idx = next_jump - num_ma_rates
+        @inbounds p.affects![idx](integrator)
     end
     return integrator.u
-end
\ No newline at end of file
+end
+
+"check if the rate is 0 and if it is, make the next jump time Inf"
+@inline function is_total_rate_zero!(p) :: Bool
+    sum_rate = p.sum_rate
+    if abs(sum_rate < eps(typeof(sum_rate)))
+        p.next_jump = 0
+        p.next_jump_time = convert(typeof(sum_rate), Inf)
+        return true
+    end
+    return false
+end
+
+"perform linear search of r over array. Output element j s.t. array[j-1] < r <= array[j]. Will error if no such r exists"
+@inline function linear_search(array, r)
+    jidx = 1
+    parsum = array[jidx]
+    while parsum < r
+        jidx   += 1
+        parsum += array[jidx]
+    end
+    return jidx
+end
+
+"perform rejection sampling test"
+@inline function rejectrx(p, u, jidx, params, t)
```

Comment:
Does passing local variables like `crhigh` and such give a difference over using the fields of `p` here?

Reply:
I think it's a tradeoff. The case, which lead to the fastest implementation is if the whole SSA is done as one big function, and fields of p are only accessed twice: at the start of the simulation and at the end. The other extreme is having a lot of little functions, making code more modular but requiring fields of p to be accessed more times. I believe in this case it is better to just access the fields of p inside the function instead of passing them around as variables, because the whole reason for the existence of p is the ability to do this. After all, wherever we pass in p, we could just pass in all the variables that p stores. This could potentially be a little bit faster but at the cost of making the code less readable and less modular. For the existing SSAs that use rejectrx, there is only a constant number of times (around 2-3) when rejectrx is called, so the difference cannot be that big. I will change it to work with local variables instead but I think it would defy the purpose of the existence of the aggregator p.

Reply:
I'm against refactoring at the expense of performance. That's why I suggested profiling the effect of this modularization on the BCR model, to see if it matters. We should use `@btime` to see if accessing the fields makes no difference, in which case we can avoid many local variables. If it does make anything more than a couple percent difference I'd say we should stick with unpacking everything before calling hot-loop methods like `rejectrx`.